### PR TITLE
Issue #286: Port read-my-writes fix from V4

### PR DESCRIPTION
**Pull request by David-Noble-at-work**
#### See https://github.com/Azure/azure-sdk-for-java/pull/6201.

From a v2.6.X perspective this PR picks up a number of diagnostic improvements that have been introduced to the Direct TCP stack. This PR also bumps the version number to 2.6.4-SNAPSHOT pending release from master.

### CODE REVIEW

File: benchmark/src/main/java/com/microsoft/azure/cosmosdb/benchmark/Main.java [hunk 1]
```diff
@@ -94,6 +94,8 @@ public static void main(String[] args) throws Exception {
             System.err.println("Invalid Usage: " + e.getMessage());
             System.err.println("Try '-help' for more information.");
             throw e;
+        } finally {
+            System.exit(0);
```
Comment (moderakh) [inline benchmark/src/main/java/com/microsoft/azure/cosmosdb/benchmark/Main.java]: There are tests which are calling Main.main(), System.exit() will cause test JVM exit.

Why is this needed?

File: changelog/README.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## 2.6.4-SNAPSHOT - 2019-11-08
+- Improved diagnostics in Direct TCP transport client ([#287](https://github.com/Azure/azure-cosmosdb-java/pull/287))
+- Bumped netty.version to 4.1.42.Final and netty-tcnative.version to 2.0.26.Final to address a Direct TCP SSL issue ([#274](https://github.com/Azure/azure-cosmosdb-java/pull/274))
+
```
Comment (moderakh) [inline changelog/README.md]: @kushagraThapar given that the changelog from github repo is imported to the Microsoft official release docs, what's the process of updating this file?

File: benchmark/src/main/resources/log4j.properties [hunk 3]
```diff
@@ -1,15 +1,29 @@
...
-log4j.category.com.microsoft.azure.cosmosdb.internal.directconnectivity.rntbd=INFO
-log4j.category.io.netty=INFO
-log4j.category.io.reactivex=INFO
-# A1 is set to be a ConsoleAppender.
-log4j.appender.A1=org.apache.log4j.ConsoleAppender
+log4j.category.com.azure.cosmos=INFO
+log4j.category.com.azure.cosmos.benchmark=INFO
+log4j.category.com.azure.cosmos.internal=INFO
+log4j.category.com.aure.cosmos.internal.caches=INFO
+log4j.category.com.aure.cosmos.internal.changefeed=INFO
```
Comment (moderakh) [inline benchmark/src/main/resources/log4j.properties]: typo: aure -> azure

File: benchmark/src/main/java/com/microsoft/azure/cosmosdb/benchmark/Main.java (hunk 1 above)
Comment (David-Noble-at-work) [reply benchmark/src/main/java/com/microsoft/azure/cosmosdb/benchmark/Main.java]: It was needed during debugging and I've now removed it because it caused a test failure.

File: changelog/README.md (hunk 2 above)
Comment (kushagraThapar) [reply changelog/README.md]: > @kushagraThapar given that the changelog from github repo is imported to the Microsoft official release docs, what's the process of updating this file?

Good point, in that case, we should change the top heading of changelog to `Unreleased` and keep it up to date with every PR. 
Once a version gets released, update the top heading to release version and date. 
The way .net does it: https://docs.microsoft.com/en-us/azure/cosmos-db/sql-api-sdk-dotnet-standard

Comment (David-Noble-at-work) [reply changelog/README.md]: I updated the changeling using the heading "## Unreleased"

File: benchmark/src/main/java/com/microsoft/azure/cosmosdb/benchmark/ReadMyWriteWorkflow.java [hunk 4]
```diff
@@ -144,8 +152,14 @@ protected void performWorkload(Subscriber<Document> subs, long i) throws Excepti
...
 
         concurrencyControlSemaphore.acquire();
+        logger.debug("concurrencyControlSemaphore: {}", concurrencyControlSemaphore);
 
-        obs.subscribeOn(Schedulers.computation()).subscribe(subs);
+        try {
+            obs.subscribeOn(Schedulers.computation()).subscribe(subs);
+        } catch (Throwable error) {
+            concurrencyControlSemaphore.release();
+            logger.error("subscription failed due to ", error);
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply benchmark/src/main/java/com/microsoft/azure/cosmosdb/benchmark/ReadMyWriteWorkflow.java]: This--I think--is a debug remnant. I would like to remove it on my follow-up PR: #289. Is that OK with you?

[the remark this answers is not in the document]
Comment (moderakh) [reply benchmark/src/main/java/com/microsoft/azure/cosmosdb/benchmark/ReadMyWriteWorkflow.java]: ok sure

**Comment by kushagraThapar** [comment]
Emulator Example tests CI passed here: https://dev.azure.com/azure-cosmos-java/azure-cosmos-java-private/_build/results?buildId=10264

**Comment by kushagraThapar** [comment]
Session Singleregion https CI passed here: https://dev.azure.com/azure-cosmos-java/azure-cosmos-java-private/_build/results?buildId=10265

**Comment by kushagraThapar** [comment]
Strong-singleregion-https-tests CI passed here: https://dev.azure.com/azure-cosmos-java/azure-cosmos-java-private/_build/results?buildId=10397

**Comment by kushagraThapar** [comment]
Merging the PR now as all the CI has passed.
